File: backend/utils/tokenGenerator.js
const getNextToken = (db, service) => {
  const prefixes = {
    Bonafide: 'B',
    Transfer: 'T',
    Fee: 'F'
  };

  if (!prefixes[service]) {
    throw new Error('invalid service');
  }

  const prefix = prefixes[service];

  // Find recent tokens for this service
  const tokens = db.findTokens({ service });
  let maxNum = 0;

  tokens.forEach(t => {
    const tn = t.token_number || '';
    try {
      const num = parseInt(tn.split('-')[1]);
      if (num > maxNum) {
        maxNum = num;
      }
    } catch (e) {
      // ignore parsing errors
    }
  });

  const nextNum = maxNum + 1;
  return `${prefix}-${nextNum}`;
};

module.exports = { getNextToken };

File: backend/utils/token_generator.py
def get_next_token(db, service: str) -> str:
    """Return next token_number for a service (prefix-number).

    Service must be one of: Bonafide, Transfer, Fee
    """
    prefixes = {'Bonafide': 'B', 'Transfer': 'T', 'Fee': 'F'}
    if service not in prefixes:
        raise ValueError('invalid service')
    prefix = prefixes[service]

    # look up recent tokens for this service and compute next numeric suffix
    docs = list(db.tokens.find({'service': service}, {'token_number': 1}).sort('created_at', -1).limit(100))
    max_num = 0
    for d in docs:
        tn = d.get('token_number', '')
        try:
            num = int(tn.split('-', 1)[1])
            if num > max_num:
                max_num = num
        except Exception:
            continue

    next_num = max_num + 1
    return f"{prefix}-{next_num}"
